jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -181,7 +181,13 @@ public HttpField getCacheControl()
 
     public void setCacheControl(HttpField cacheControl)
     {
-        _cacheControl = cacheControl;
+        if (cacheControl == null)
+            _cacheControl = null;
+        if (cacheControl.getHeader() != HttpHeader.CACHE_CONTROL)
+            throw new IllegalArgumentException("!Cache-Control");
+        _cacheControl = cacheControl instanceof PreEncodedHttpField
+            ? cacheControl
+            : new PreEncodedHttpField(cacheControl.getHeader(), cacheControl.getValue());
```

Comment:
Using a preencoded field here is worthwhile as we can pre-compute the bytes used for both HTTP/1 and HTTP/2 hpack, so the field can be added with a single byte array copy rather than converting 2 strings to bytes each time